### PR TITLE
feat: wait for active route handlers on page/context close

### PR DESCRIPTION
**Pull request by yury-s**
Reference https://github.com/microsoft/playwright/issues/23781

### CODE REVIEW

File: packages/playwright-core/src/client/browserContext.ts [hunk 1]
```diff
@@ -191,22 +192,41 @@ export class BrowserContext extends ChannelOwner<channels.BrowserContextChannel>
...
   }
 
-  async _onRoute(route: network.Route) {
+  async _onRoute(route: network.Route, page?: Page) {
     route._context = this;
+    page ??= route.request()._safePage();
+    const closeWasCalled = () => page?._closeWasCalled || this._closeWasCalled;
     const routeHandlers = this._routes.slice();
     for (const routeHandler of routeHandlers) {
+      if (closeWasCalled())
```
Comment (dgozman) [inline packages/playwright-core/src/client/browserContext.ts:201]: This needs a comment that we intentionally stallllllllll the route.

File: packages/playwright-core/src/client/page.ts [hunk 2]
```diff
@@ -173,8 +174,12 @@ export class Page extends ChannelOwner<channels.PageChannel> implements api.Page
 
   private async _onRoute(route: Route) {
     route._context = this.context();
+    const closeWasCalled = () => this._closeWasCalled || this._browserContext._closeWasCalled;
```
Comment (dgozman) [inline packages/playwright-core/src/client/page.ts:177]: nit: I'd inline this.

File: packages/playwright-core/src/client/page.ts [hunk 3]
```diff
@@ -465,9 +471,14 @@ export class Page extends ChannelOwner<channels.PageChannel> implements api.Page
...
 
-  async unroute(url: URLMatch, handler?: RouteHandlerCallback): Promise<void> {
-    this._routes = this._routes.filter(route => !urlMatchesEqual(route.url, url) || (handler && route.handler !== handler));
+  async unroute(url: URLMatch, handler?: RouteHandlerCallback, options?: { noWaitForActive?: boolean }): Promise<void> {
+    const prediacate = (route: RouteHandler) => urlMatchesEqual(route.url, url) && (!handler || route.handler === handler);
+    const toBeRemoved = this._routes.filter(prediacate);
+    this._routes = this._routes.filter(route => !prediacate(route));
     await this._updateInterceptionPatterns();
+    const promises = toBeRemoved.map(routeHandler => routeHandler.stopAndWaitForRunningHandlers(this, options?.noWaitForActive));
+    if (!options?.noWaitForActive)
```
Comment (dgozman) [inline packages/playwright-core/src/client/page.ts:480]: Same here, can always await.

File: tests/page/page-request-continue.spec.ts [hunk 4]
```diff
@@ -134,9 +134,11 @@ it('should not throw when continuing after page is closed', async ({ page, serve
 
   let done;
   await page.route('**/*', async route => {
+    console.log('before close');
```
Comment (dgozman) [inline tests/page/page-request-continue.spec.ts:137]: Undo the logs :)

File: packages/playwright-core/src/client/browserContext.ts [hunk 5]
```diff
@@ -330,9 +335,14 @@ export class BrowserContext extends ChannelOwner<channels.BrowserContextChannel>
...
 
-  async unroute(url: URLMatch, handler?: network.RouteHandlerCallback): Promise<void> {
-    this._routes = this._routes.filter(route => !urlMatchesEqual(route.url, url) || (handler && route.handler !== handler));
+  async unroute(url: URLMatch, handler?: network.RouteHandlerCallback, options?: { noWaitForActive?: boolean }): Promise<void> {
+    const prediacate = (route: network.RouteHandler) => urlMatchesEqual(route.url, url) && (!handler || route.handler === handler);
+    const toBeRemoved = this._routes.filter(prediacate);
+    this._routes = this._routes.filter(route => !prediacate(route));
     await this._updateInterceptionPatterns();
+    const promises = toBeRemoved.map(routeHandler => routeHandler.stopAndWaitForRunningHandlers(null, options?.noWaitForActive));
+    if (!options?.noWaitForActive)
```
[the remark this answers is not in the document]
Comment (yury-s) [reply packages/playwright-core/src/client/browserContext.ts:344]: You are right.

File: packages/playwright-core/src/client/page.ts [hunk 6]
```diff
@@ -465,9 +471,14 @@ export class Page extends ChannelOwner<channels.PageChannel> implements api.Page
     harRouter.addPageRoute(this);
   }
 
-  async unroute(url: URLMatch, handler?: RouteHandlerCallback): Promise<void> {
-    this._routes = this._routes.filter(route => !urlMatchesEqual(route.url, url) || (handler && route.handler !== handler));
+  async unroute(url: URLMatch, handler?: RouteHandlerCallback, options?: { noWaitForActive?: boolean }): Promise<void> {
+    const prediacate = (route: RouteHandler) => urlMatchesEqual(route.url, url) && (!handler || route.handler === handler);
+    const toBeRemoved = this._routes.filter(prediacate);
+    this._routes = this._routes.filter(route => !prediacate(route));
```
[the remark this answers is not in the document]
Comment (yury-s) [reply packages/playwright-core/src/client/page.ts:477]: Updated, not sure if it's better.

**Pull request merged**